pkg/assembler/backends/ent/backend/package.go

```diff
@@ -0,0 +1,377 @@
+	if err != nil {
+		return nil, errors.Wrap(err, "upsert package namespace")
+	}
+
+	nameID, err := client.PackageName.Create().SetNamespaceID(nsID).SetName(pkg.Name).
+		OnConflict(sql.ConflictColumns(packagename.FieldName, packagename.FieldNamespaceID)).UpdateNewValues().ID(ctx)
+	if err != nil {
+		return nil, errors.Wrap(err, "upsert package name")
+	}
+
+	pvID, err := client.PackageVersion.Create().
+		SetNameID(nameID).
+		SetVersion(valueOrDefault(pkg.Version, "")).
+		SetSubpath(valueOrDefault(pkg.Subpath, "")).
+		SetQualifiers(normalizeInputQualifiers(pkg.Qualifiers)).
+		SetHash(versionHashFromInputSpec(pkg)).
+		OnConflict(
+			sql.ConflictColumns(
+				packageversion.FieldHash,
+				packageversion.FieldNameID,
+			),
+		).
+		UpdateNewValues().
+		ID(ctx)
+	if err != nil {
+		return nil, errors.Wrap(err, "upsert package version")
+	}
+
+	pv, err := client.PackageVersion.Query().Where(packageversion.IDEQ(pvID)).
+		WithName(withPackageNameTree()).
+		Only(ctx)
+	if err != nil {
+		return nil, errors.Wrap(err, "get package version")
+	}
+
+	return pv, nil
+}
+
+func withPackageVersionTree() func(*ent.PackageVersionQuery) {
+	return func(q *ent.PackageVersionQuery) {
+		q.WithName(withPackageNameTree())
+	}
+}
+
+func withPackageNameTree() func(q *ent.PackageNameQuery) {
+	return func(q *ent.PackageNameQuery) {
+		q.WithNamespace(func(q *ent.PackageNamespaceQuery) {
+			q.WithPackage()
+		})
+	}
+}
+
+func versionHashFromInputSpec(pkg model.PkgInputSpec) string {
+	return hashPackageVersion(
+		valueOrDefault(pkg.Version, ""),
+		valueOrDefault(pkg.Subpath, ""),
+		normalizeInputQualifiers(pkg.Qualifiers))
+}
+
+func hashPackageVersion(version, subpath string, qualifiers []model.PackageQualifier) string {
+	hash := sha1.New()
+	hash.Write([]byte(version))
+	hash.Write([]byte(subpath))
+	qualifiersBuffer := bytes.NewBuffer(nil)
+
+	sort.Slice(qualifiers, func(i, j int) bool { return qualifiers[i].Key < qualifiers[j].Key })
+
+	for _, qualifier := range qualifiers {
+		qualifiersBuffer.WriteString(qualifier.Key)
+		qualifiersBuffer.WriteString(qualifier.Value)
+	}
+
+	hash.Write(qualifiersBuffer.Bytes())
+	return fmt.Sprintf("%x", hash.Sum(nil))
+}
+
+func normalizeInputQualifiers(inputs []*model.PackageQualifierInputSpec) []model.PackageQualifier {
+	qualifiers := []model.PackageQualifier{}
+	for _, q := range inputs {
+		qualifiers = append(qualifiers, model.PackageQualifier{
+			Key:   q.Key,
+			Value: q.Value,
+		})
+	}
+
+	return qualifiers
+}
+
+func qualifiersToSpecQualifiers(q []*model.PackageQualifierInputSpec) []*model.PackageQualifierSpec {
+	results := make([]*model.PackageQualifierSpec, len(q))
+	for i, s := range q {
+		results[i] = &model.PackageQualifierSpec{
+			Key:   s.Key,
+			Value: &s.Value,
+		}
+	}
+	return results
+}
+
+func packageVersionQuery(spec *model.PkgInputSpec) predicate.PackageVersion {
```

Comment:
I believe these `where` predicates are failing to match two package versions that are identical but which have different qualifiers.

Running this input against this data using `getPkgVersion` fn will result in an error as it expects only one result, but will return two

Input:
```
namespace := "debian"
version := "3.11.2-1+b1"
subpath := ""
pkgin := model.PkgInputSpec{
		Type:       "deb",
		Namespace:  &namespace,
		Name:       "python3",
		Version:    &version,
		Qualifiers: []*model.PackageQualifierInputSpec{},
		Subpath:    &subpath,
	}
```

DB: 
```
[
	{
		"id": 64424511484,
		"version": "3.11.2-1+b1",
		"subpath": "",
		"qualifiers": [
			{
				"key": "arch",
				"value": "arm64"
			},
			{
				"key": "distro",
				"value": "debian-12"
			},
			{
				"key": "upstream",
				"value": "python3-defaults@3.11.2-1"
			}
		],
		"hash": "b1a934efb143f8d71e0d342c9f23066124d36f7e",
		"name_id": 51539609596,
		"tenant_id": 90194313216
	},
	{
		"id": 64424512478,
		"version": "3.11.2-1+b1",
		"subpath": "",
		"qualifiers": [
		],
		"hash": "14a747debb4ca0ca8b878d0e4913720e8a0e8b70",
		"name_id": 51539609596,
		"tenant_id": 90194313216
	}
]
```